File: formula-py/formula/shunting_yard.py
import enum
import string
from typing import Callable, List, Self, TypeAlias

from formula.data_context import DataContext
from formula.resolvable import Resolvable
from formula.resolve_error import ResolveError
from formula.resolved_value import ResolvedValue, QuotedTextResolvedValue
from formula.token import token_tree

_TokenMapper: TypeAlias = Callable[[list[ResolvedValue]], ResolvedValue]
_DataResolver: TypeAlias = Callable[[DataContext], ResolvedValue]
_VariableResolver: TypeAlias = Callable[[DataContext, str], ResolvedValue]


class Associativity(enum.IntEnum):
    LEFT = 1
    RIGHT = 2


class Node:
    pass


class OperatorFunction(Node):
    def __init__(self, name: str, operands: int | None, fn: _TokenMapper):
        super().__init__()
        self.name = name
        self.operands = operands
        self.fn = fn

    def execute(self, args: list[ResolvedValue]):
        if self.operands is not None and self.operands != len(args):
            raise Exception(f"Invalid number of operands; expected {self.operands} but got {len(args)}")
        return self.fn(args)

    def __repr__(self):
        return self.name


class Operator(OperatorFunction):
    def __init__(self,
                 name: str,
                 operands: int,
                 precedence: int,
                 associativity: Associativity,
                 fn: _TokenMapper):
        super().__init__(name, operands, fn)
        self.precedence = precedence
        self.associativity = associativity


class BiOperator(Node):
    def __init__(self, unary: Operator, binary: Operator):
        super().__init__()
        self.unary = unary
        self.binary = binary


class Function(OperatorFunction):
    def __init__(self, name: str, operands: int, fn: _TokenMapper):
        super().__init__(name, operands, fn)


class VarargsFunction(OperatorFunction):
    def __init__(self, name: str, fn: _TokenMapper):
        super().__init__(name, None, fn)


class Variable(Node):
    def __init__(self, key: str, resolve_fn: _VariableResolver):
        super().__init__()
        self.key = key
        self.resolve_fn = resolve_fn

    def resolve(self, context: DataContext) -> ResolvedValue:
        return self.resolve_fn(context, self.key)

    def __str__(self):
        return self.key


class Comment(Node):
    def __init__(self, text: str, decorate_fn: Callable[[str, ResolvedValue], ResolvedValue]):
        super().__init__()
        self.text = text
        self.decorate_fn = decorate_fn

    def apply(self, previous: ResolvedValue):
        return self.decorate_fn(self.text, previous)


class Term(Node):
    def __init__(self, resolve_fn: Callable[[], ResolvedValue], prefix: str | None = None, suffix: str | None = None):
        super().__init__()
        self.prefix = prefix
        self.suffix = suffix
        self.resolve_fn = resolve_fn

    def resolve(self) -> ResolvedValue:
        resolved = self.resolve_fn()
        if self.prefix is not None and self.suffix is not None:
            return QuotedTextResolvedValue(resolved, self.prefix, self.suffix)
        return resolved

    def __repr__(self):
        return str(self.resolve())


class ShuntingYard:
    def __init__(self, stack: List[Node | str | int | float | None], original_formula: str):
        self._stack = stack
        self.original_formula = original_formula

    def resolve(self, context: DataContext | None):
        local_stack = list[any]()
        stack = list[any]()
        stack += self._stack

        while len(stack) > 0:
            next_token = stack.pop(0)
            if isinstance(next_token, OperatorFunction):
                operand_count = next_token.operands \
                    if next_token.operands is not None \
                    else self._pop_arity(local_stack, next_token)
                parameters = []
                for i in range(operand_count):
                    parameters.insert(0, self._checked_popped_param(next_token, i, local_stack))
                local_stack.insert(0, next_token.execute(parameters))
            elif isinstance(next_token, Term):
                stack.insert(0, next_token.resolve())
            elif isinstance(next_token, Variable):
                stack.insert(0, next_token.resolve(context))
            elif isinstance(next_token, Comment):
                local_stack.insert(0, next_token.apply(local_stack.pop(0)))
            else:
                if isinstance(next_token, Resolvable):
                    local_stack.insert(0, next_token.resolve(context))
                else:
                    local_stack.insert(0, next_token)

        return local_stack.pop(0) if len(local_stack) > 0 else ResolvedValue(None)

    @staticmethod
    def _checked_popped_param(func: any, index: int, stack: list):
        if len(stack) == 0:
            raise ResolveError(f"Missing parameter #{index+1} for \"{func}\"")
        return stack.pop(0)

    @staticmethod
    def _pop_arity(stack: list, func: OperatorFunction):
        if len(stack) == 0:
            raise ResolveError(f"Missing arity count for \"{func}\"")
        return stack.pop(0)


class ShuntingYardParser:
    def __init__(self):
        self._parser = (token_tree.create()
                        .ignore_whitespace()
                        .add_branch(token_tree.DECIMAL, lambda x: float(x))
                        .add_branch(token_tree.INTEGER, lambda x: int(x))
                        .add_branch('(', lambda x: x)
                        .add_branch(')', lambda x: x)
                        .add_branch(',', lambda x: x)
                        .add_branch(token_tree.literal('"', '"', '\\"'),
                                    lambda quote: Term(lambda: ResolvedValue(quote[1:-1]), '"', '"'))
                        .add_branch(token_tree.literal('\'', '\'', '\\\''),
                                    lambda quote: Term(lambda: ResolvedValue(quote[1:-1]), '\'', '\'')))

    def operator(self, symbol: str,
                 precedence: int,
                 associativity: Associativity,
                 operands: int,
                 fn: _TokenMapper) -> Self:
        self._parser.add_branch(symbol, lambda _: Operator(symbol, operands, precedence, associativity, fn))
        return self

    def decimal_operator_2(self,
                           symbol: str,
                           precedence: int,
                           associativity: Associativity,
                           func: Callable[[float, float], str | int | float | bool]) -> Self:
        self._parser.add_branch(symbol,
                                lambda _: Operator(symbol,
                                                   2,
                                                   precedence,
                                                   associativity,
                                                   lambda p: ResolvedValue(func(p[0].as_decimal(), p[1].as_decimal()))))
        return self

    def decimal_bi_operator(self,
                            symbol: str,
                            unary_precedence: int,
                            unary_associativity: Associativity,
                            unary_func: Callable[[float], any],
                            binary_precedence: int,
                            binary_associativity: Associativity,
                            binary_func: Callable[[float, float], any]) -> Self:
        self._parser.add_branch(token_tree.term(symbol), lambda _: BiOperator(
            Operator(symbol, 1, unary_precedence, unary_associativity,
                     lambda p: ResolvedValue(unary_func(p[0].as_decimal()))),
            Operator(symbol, 2, binary_precedence, binary_associativity,
                     lambda p: ResolvedValue(binary_func(p[0].as_decimal(), p[1].as_decimal())))
        ))
        return self

    def term(self, text: str, extractor: Callable[[], ResolvedValue]) -> Self:
        self._parser.add_branch(token_tree.term(text), lambda _: Term(extractor))
        return self

    def function_1(self, name, func: Callable[[ResolvedValue], ResolvedValue]) -> Self:
        self._parser.add_branch(token_tree.term(name), lambda _: Function(name, 1, lambda x: func(x[0])))
        return self

    def function_2(self, name, func: Callable[[ResolvedValue, ResolvedValue], ResolvedValue]) -> Self:
        self._parser.add_branch(token_tree.term(name), lambda _: Function(name, 2, lambda x: func(x[0], x[1])))
        return self

    def function_n(self, name, func: Callable[[list[ResolvedValue]], ResolvedValue]) -> Self:
        self._parser.add_branch(token_tree.term(name), lambda _: VarargsFunction(name, func))
        return self

    def variable(self, prefix: str, suffix: str, func: Callable[[DataContext, str], ResolvedValue]) -> Self:
        self._parser.add_branch(
            [
                *token_tree.term(prefix),
                token_tree.any_of(string.ascii_letters),
                token_tree.optional(token_tree.KEY),
                *token_tree.term(suffix)
            ],
            lambda key: Variable(key, lambda context, v_key: func(context, key[len(prefix):len(v_key) - len(suffix)])))
        return self

    def comment(self, prefix: str, suffix: str, func: Callable[[str, ResolvedValue], ResolvedValue]) -> Self:
        self._parser.add_branch(token_tree.literal(prefix, suffix),
                                lambda token: Comment(token[len(prefix):len(token) - len(suffix)], func))
        return self

    def decimal_function_1(self, name, func: Callable[[float], any]) -> Self:
        self._parser.add_branch(token_tree.term(name),
                                lambda _: Function(name, 1,
                                                   lambda x: ResolvedValue(func(x[0].as_decimal()))))
        return self

    def decimal_function_2(self, name, func: Callable[[float, float], any]) -> Self:
        self._parser.add_branch(token_tree.term(name),
                                lambda _: Function(name, 2,
                                                   lambda x: ResolvedValue(func(x[0].as_decimal(), x[1].as_decimal()))))
        return self

    def function_3(self, name, func: Callable[[ResolvedValue, ResolvedValue, ResolvedValue], ResolvedValue]):
        self._parser.add_branch(token_tree.term(name),
                                lambda _: Function(name, 3,
                                                   lambda x: func(x[0], x[1], x[2])))
        return self

    def parse(self, formula: str) -> ShuntingYard:
        output_buffer = list()
        operator_stack = list()
        arity_stack = list()
        tokens = self._parser.parse(formula)

        for i in range(len(tokens)):
            token = tokens[i]
            previous = tokens[i - 1] if i > 0 else None

            if isinstance(token, BiOperator):
                operator = token
                token = operator.binary
                if previous is None or isinstance(previous, Operator) or previous == '(' or previous == ',':
                    token = operator.unary

            if isinstance(token, Operator):
                if len(operator_stack) > 0:
                    top = operator_stack[-1]
                    if isinstance(top, Operator):
                        if (token.precedence < top.precedence
                                or (token.associativity == Associativity.LEFT and token.precedence == top.precedence)):
                            operator_stack.pop()
                            output_buffer.append(top)
                operator_stack.append(token)
                continue

            if isinstance(token, Function) or isinstance(token, VarargsFunction):
                operator_stack.append(token)
                arity_stack.append(1)
                continue

            if isinstance(token, Variable) or isinstance(token, Term) or isinstance(token, Comment):
                output_buffer.append(token)
                continue

            match token:
                case ' ' | '{' | '}':
                    pass
                case ',':
                    arity_stack[-1] += 1
                    while len(operator_stack) > 0:
                        top = operator_stack.pop()
                        if top == '(':
                            operator_stack.append(top)
                            break
                        output_buffer.append(top)
                case '(':
                    operator_stack.append(token)
                case ')':
                    while len(operator_stack) > 0:
                        top = operator_stack.pop()
                        if top == '(':
                            break
                        output_buffer.append(top)
                    if len(operator_stack) > 0:
                        if isinstance(operator_stack[-1], Function):
                            output_buffer.append(operator_stack.pop())
                        elif isinstance(operator_stack[-1], VarargsFunction):
                            arity = 0 if previous == '(' else arity_stack.pop()
                            output_buffer.append(arity)
                            output_buffer.append(operator_stack.pop())
                case _:
                    output_buffer.append(ResolvedValue(token))

        while len(operator_stack) > 0:
            output_buffer.append(operator_stack.pop())

        return ShuntingYard(output_buffer, formula)

File: js/formula-js/src/ShuntingYard.ts
import {DataContext} from "./DataContext";
import Parser from "./Parser";
import {QuotedTextResolvedValue} from "./QuotedTextResolvedValue";
import {Resolvable} from "./Resolvable";
import {ResolvedValue} from "./ResolvedValue";
import {ResolveError} from "./ResolveError";
import TokenTree, {alpha, decimal, integer, key, literal, optional, term} from "./TokenTree";

type ZeroOperandFunction<T> = () => T;
type OneOperandFunction<T> = (x: T) => T;
type TwoOperandFunction<T> = (x: T, y: T) => T;
type ThreeOperandFunction<T> = (x: T, y: T, z: T) => T;
type OperandFunction<T> =
    ZeroOperandFunction<T>
    | OneOperandFunction<T>
    | TwoOperandFunction<T>
    | ThreeOperandFunction<T>;
type VarargsOperandFunction<T> = (n: T[]) => T;

function mapIntFunction(operands: number, fn: OperandFunction<number>): OperandFunction<ResolvedValue> {
  switch (operands) {
    case 0:
      return () => ResolvedValue.of((fn as ZeroOperandFunction<number>)());
    case 1:
      return (x?: ResolvedValue) => ResolvedValue.of((fn as OneOperandFunction<number>)(x?.asNumber() ?? 0));
    case 2:
      return (x?: ResolvedValue, y?: ResolvedValue) => ResolvedValue.of((fn as TwoOperandFunction<number>)(x?.asNumber() ?? 0, y?.asNumber() ?? 0));
    case 3:
      return (x?: ResolvedValue, y?: ResolvedValue, z?: ResolvedValue) => ResolvedValue.of((fn as ThreeOperandFunction<number>)(x?.asNumber() ?? 0, y?.asNumber() ?? 0, z?.asNumber() ?? 0));
    default:
      throw new Error(`Unsupported number of operands: ${operands}`);
  }
}

export enum Associativity {
  Left = 1,
  Right = 2
}

interface Node {

}

export abstract class Modifier {
  abstract apply(value: ResolvedValue): ResolvedValue;
}

class NoModifier extends Modifier {
  static readonly instance = new NoModifier();

  apply(value: ResolvedValue): ResolvedValue {
    return value;
  }
}

abstract class OperatorFunction implements Node {
  public readonly name: string;
  public readonly operands: number;
  private readonly fn: OperandFunction<ResolvedValue>;

  protected constructor(name: string, operands: number, fn: OperandFunction<ResolvedValue>) {
    this.name = name;
    this.operands = operands;
    this.fn = fn;
  }

  execute(modifier: Modifier, x?: ResolvedValue, y?: ResolvedValue, z?: ResolvedValue): ResolvedValue {
    const modifiedX = modifier.apply(x);
    const modifiedY = modifier.apply(y);
    const modifiedZ = modifier.apply(z);

    switch (this.operands) {
      case 0:
        return (this.fn as ZeroOperandFunction<ResolvedValue>)();
      case 1:
        return (this.fn as OneOperandFunction<ResolvedValue>)(modifiedX);
      case 2:
        return (this.fn as TwoOperandFunction<ResolvedValue>)(modifiedX, modifiedY);
      case 3:
        return (this.fn as ThreeOperandFunction<ResolvedValue>)(modifiedX, modifiedY, modifiedZ);
      default:
        throw new Error(`Unsupported number of operands: ${this.operands}`);
    }
  }
}

class Function extends OperatorFunction {
  constructor(name: string, operands: number, fn: OperandFunction<ResolvedValue>) {
    super(name, operands, fn);
  }
}

class VarargsFunction implements Node {
  constructor(public readonly name: string,
              private readonly fn: VarargsOperandFunction<ResolvedValue>) {
  }

  execute(modifier: Modifier, args: ResolvedValue[]): ResolvedValue {
    return this.fn(args.map(modifier.apply));
  }
}

class Operator extends OperatorFunction {
  constructor(public readonly name: string,
              public readonly precedence: number,
              public readonly associativity: Associativity,
              operands: number, fn: OperandFunction<ResolvedValue>) {
    super(name, operands, fn);
  }
}

class BiOperator implements Node {
  constructor(public readonly unary: Operator, public readonly binary: Operator) {
  }
}

class Variable implements Node {
  constructor(public readonly key: string,
              private readonly resolver: (context: DataContext, key: string) => ResolvedValue | Resolvable | undefined) {
  }

  resolve(context: DataContext): ResolvedValue | Resolvable {
    return this.resolver(context, this.key) ?? ResolvedValue.None;
  }

  public toString(): string {
    return this.key;
  }
}

class Comment implements Node {
  constructor(private readonly text: string,
              private readonly decorator: (text: string, value: ResolvedValue) => ResolvedValue) {
  }

  apply(previous: ResolvedValue) {
    return this.decorator(this.text, previous);
  }
}

class Term implements Node {
  constructor(private readonly resolver: (context: DataContext) => ResolvedValue | Resolvable | undefined,
              private readonly prefix: string | undefined = undefined,
              private readonly suffix: string | undefined = undefined) {
  }

  resolve(context: DataContext) {
    return (this.prefix && this.suffix)
        ? this.resolver(context).map(resolved => new QuotedTextResolvedValue(resolved, this.prefix, this.suffix))
        : this.resolver(context);
  }
}

export class ShuntingYardParser implements Parser {
  private readonly parser: TokenTree;
  private bracketFn: (value: ResolvedValue) => ResolvedValue = value => value;

  constructor() {
    this.parser = new TokenTree()
    .ignoreWhitespaces()
    .add([integer], token => parseInt(token))
    .add([decimal], token => parseFloat(token))
    .add('(', token => token)
    .add(')', token => token)
    .add(';', token => token)
    .add(literal('"', '"', '\\"'), quote => new Term(() => ResolvedValue.of(quote.slice(1, -1)), "\"", "\""))
    .add(literal("'", "'", "\\'"), quote => new Term(() => ResolvedValue.of(quote.slice(1, -1)), "'", "'"))
  }

  operator(symbol: string, precedence: number, associativity: Associativity, operands: number, fn: OperandFunction<ResolvedValue>) {
    this.parser.add(symbol, _ => new Operator(symbol, precedence, associativity, operands, fn));
    return this;
  }

  biOperator(symbol: string,
             unaryPrecedence: number, unaryAssociativity: Associativity, unaryFn: OneOperandFunction<ResolvedValue>,
             binaryPrecedence: number, binaryAssociativity: Associativity, binaryFn: TwoOperandFunction<ResolvedValue>) {
    this.parser.add(symbol, _ => new BiOperator(
        new Operator(symbol, unaryPrecedence, unaryAssociativity, 1, unaryFn),
        new Operator(symbol, binaryPrecedence, binaryAssociativity, 2, binaryFn)
    ));
    return this;
  }

  intOperator(symbol: string, precedence: number, associativity: Associativity, operands: number, fn: OperandFunction<number>) {
    return this.operator(symbol, precedence, associativity, operands, mapIntFunction(operands, fn));
  }

  function(name: string, operands: number, fn: OperandFunction<ResolvedValue>) {
    this.parser.add(name, _ => new Function(name, operands, fn));
    return this;
  }

  modifier(name: string, mod: Modifier) {
    this.parser.add(name, _ => mod);
    return this;
  }

  varargsFunction(name: string, fn: VarargsOperandFunction<ResolvedValue>) {
    this.parser.add(name, _ => new VarargsFunction(name, fn));
    return this;
  }

  intFunction(name: string, operands: number, fn: OperandFunction<number>) {
    return this.function(name, operands, mapIntFunction(operands, fn));
  }

  variable(prefix: string, suffix: string, extractor: (context: DataContext, key: string) => ResolvedValue | Resolvable) {
    this.parser.add([term(prefix), alpha, optional(key), term(suffix)],
        key => new Variable(key, (context: DataContext, key: string) =>
            extractor(context, key.substring(prefix.length, key.length - suffix.length))));
    return this;
  }

  comment(prefix: string, suffix: string, decorator: (text: string, value: ResolvedValue) => ResolvedValue = (text, value) => value) {
    this.parser.add(literal(prefix, suffix),
        key => new Comment(key.substring(prefix.length, key.length - suffix.length), decorator));
    return this;
  }

  term(text: string, extractor: (context: DataContext) => ResolvedValue | Resolvable | undefined) {
    this.parser.add([term(text)],
        key => new Variable(key, (context: DataContext) =>
            extractor(context)));
    return this;
  }

  brackets(mapFn: (value: ResolvedValue) => ResolvedValue) {
    this.bracketFn = mapFn;
    return this;
  }

  parse(formula: string): ShuntingYard {
    let outputBuffer: OperatorStack = [];
    let operatorStack: OperatorStack = [];
    let arityStack: number[] = [];

    const tokens = this.parser.parse(formula);

    for (let i = 0; i < tokens.length; i++) {
      let token = tokens[i];
      let previous = i > 0 ? tokens[i - 1] : undefined;

      if (token instanceof BiOperator) {
        let operator = token;
        token = operator.binary;
        if (!previous || previous instanceof Operator || previous === '(' || previous === ';') {
          token = operator.unary;
        }
      }

      if (token instanceof Operator) {
        let operator = token;
        let top = operatorStack[operatorStack.length-1];
        if (top instanceof Operator) {
          if ((operator.precedence < top.precedence)
              || (operator.associativity === Associativity.Left
                  && operator.precedence === top.precedence)) {
            operatorStack.pop();
            outputBuffer.push(top);
          }
        }

        operatorStack.push(operator);
        continue;
      }

      if (token instanceof Modifier) {
        operatorStack.push(token);
        continue;
      }

      if (token instanceof Function) {
        operatorStack.push(token);
        arityStack.push(1);
        continue;
      }

      if (token instanceof VarargsFunction) {
        operatorStack.push(token);
        arityStack.push(1);
        continue;
      }

      if (token instanceof Variable) {
        outputBuffer.push(token);
        continue;
      }

      if (token instanceof Term) {
        outputBuffer.push(token);
        continue;
      }

      if (token instanceof Comment) {
        outputBuffer.push(token);
        continue;
      }

      switch (token) {
        case ' ':
        case '{':
        case '}':
          // ignore
          break;
        case ';':
          arityStack[arityStack.length - 1]++;
          while (operatorStack.length > 0) {
            let next = operatorStack.pop();
            if (next === '(') {
              operatorStack.push(next);
              break;
            }
            outputBuffer.push(next);
          }
          break;
        case '(':
          operatorStack.push(token);
          break;
        case ')':
          while (operatorStack.length > 0) {
            let next = operatorStack.pop();
            if (next === '(') {
              break;
            }
            outputBuffer.push(next);
          }

          if (operatorStack[operatorStack.length-1] instanceof Function) {
            const paramCount = arityStack.pop();
            const func = operatorStack.pop() as Function;
            if (paramCount !== func.operands) {
              throw new Error(`${func.name} expected ${func.operands} parameters, but got ${paramCount}`);
            }
            outputBuffer.push(func);
          } else if (operatorStack[operatorStack.length-1] instanceof VarargsFunction) {
            let arity = arityStack.pop();
            arity = previous === '(' ? 0 : arity;
            outputBuffer.push(arity);
            outputBuffer.push(operatorStack.pop());
          } else if (operatorStack[operatorStack.length-1] instanceof Modifier) {
            outputBuffer.push(operatorStack.pop() as Modifier);
          }
          break;
        default:
          outputBuffer.push(ResolvedValue.of(token));
      }
    }

    while (operatorStack.length > 0) {
      outputBuffer.push(operatorStack.pop());
    }

    return new ShuntingYard(outputBuffer, formula);
  }
}

type OperatorStack = Array<Node | string | number | null | undefined>;

export class ShuntingYard extends Resolvable {
  private readonly originalFormula: string;
  private readonly stack: OperatorStack;

  static parser(): ShuntingYardParser {
    return new ShuntingYardParser();
  }

  constructor(stack: OperatorStack, originalFormula: string) {
    super();
    this.stack = stack;
    this.originalFormula = originalFormula;
  }

  asFormula(): string {
    return this.originalFormula;
  }

  resolve(context: DataContext = DataContext.Empty): ResolvedValue | undefined {
    let localStack: OperatorStack = [];
    let stack: OperatorStack = [...this.stack];
    let modifierStack: Modifier[] = [NoModifier.instance];
    while (stack.length > 0) {
      let next = stack.shift();

      if (next instanceof Modifier) {
        modifierStack.push(next);
        continue;
      }

      if (next instanceof OperatorFunction) {
        let func = next;
        if (func.operands === 0) {
          localStack.push(func.execute(modifierStack[modifierStack.length-1]));
        } else if (func.operands === 1) {
          let x = localStack.pop() as ResolvedValue;
          if (x === undefined) throw new ResolveError(`Missing parameter #1 for "${func.name}"`);
          localStack.push(func.execute(modifierStack[modifierStack.length-1], x));
        } else if (func.operands === 2) {
          let b = localStack.pop() as ResolvedValue;
          let a = localStack.pop() as ResolvedValue;
          if (b === undefined) throw new ResolveError(`Missing parameter #1 for "${func.name}"`)
          if (a === undefined) throw new ResolveError(`Missing parameter #2 for "${func.name}"`)
          localStack.push(func.execute(modifierStack[modifierStack.length-1], a, b));
        } else if (func.operands === 3) {
          let c = localStack.pop() as ResolvedValue;
          let b = localStack.pop() as ResolvedValue;
          let a = localStack.pop() as ResolvedValue;
          if (c === undefined) throw new ResolveError(`Missing parameter #1 for "${func.name}"`)
          if (b === undefined) throw new ResolveError(`Missing parameter #2 for "${func.name}"`)
          if (a === undefined) throw new ResolveError(`Missing parameter #3 for "${func.name}"`)
          localStack.push(func.execute(modifierStack[modifierStack.length-1], a, b, c));
        } else {
          throw new Error("Unsupported number of operands: " + func.operands);
        }
        continue;
      }

      if (next instanceof VarargsFunction) {
        let func = next;
        let params: ResolvedValue[] = [];
        let paramCount = localStack.pop() as number;
        while (paramCount-- > 0) {
          params.unshift(localStack.pop() as ResolvedValue);
        }
        const modifier = modifierStack[modifierStack.length-1];
        localStack.push(func.execute(modifier, params));
        continue;
      }

      if (next instanceof Comment) {
        const previous = localStack.pop();
        localStack.push(next.apply(previous as ResolvedValue));
        continue;
      }

      if (next instanceof Term) {
        next = next.resolve(context);
      }

      if (next instanceof Variable) {
        next = next.resolve(context);
        if (next instanceof ShuntingYard) {
          const nextStack = next.stack;
          for (let i = nextStack.length - 1; i >= 0; i--) {
            stack.unshift(nextStack[i]);
          }
          continue;
        }
      }

      while (next instanceof Resolvable) {
        next = next.resolve(context);
      }

      localStack.push(next);
    }

    return modifierStack[modifierStack.length-1].apply(localStack.pop() as ResolvedValue);
  }
}